Add "netstandard" moniker

joelverhagen opened the pull request:
Per NuGet/Home#1691.

Issues I need some guidance on.
- Is the unit testing I have added sufficient? There seems to be a ton of tests around the old `dotnet` moniker. I tried to replicate or retrofit them for `netstandard` as appropriate. However, there does not seem to be a consistent approach in the `CompatibilityTests.cs` suite.
- What sort of manual verification should I perform? All of the unit tests in the solution passed (those run within Visual Studio).
- Should the [design document](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/standard-platform.md) be updated to match the state of the code? I have spoke to @ericstj offline about some things and tried to meet the latest requirements.

joelverhagen committed:
Implement netstandard moniker and establish precedence over dotnet

dnfclas commented:
Hi **@joelverhagen**, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

joelverhagen commented:
/cc @emgarten @yishaigalatzer

davidfowl commented:
Don't remove any of the dotnet stuff, at least not yet. We need to keep it around for now.

joelverhagen commented:
I tried to keep existing `dotnet` functionality by not removing `dotnet` and adding `netstandardA.B <-> dotnetX.Y` equivalency mappings for each of the pairs. Forgive me if I am missing something here -- I am pretty new to this code!

emgarten commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -509,22 +532,7 @@ public void Compatibility_ProjectCanInstallDotNetLibraries(string framework)
-        [Theory]
-        [InlineData("dotnet")]
-        [InlineData("dotnet50")]
-        public void Compatibility_DotNetProjectCompat(string framework)
```
This test needs to stay

emgarten commented:
:watch:  This change should be adding netstandard, not removing any of the dotnet stuff. All of those tests need to stay. You can add new tests and theories that are similar to the dotnet tests, but it's critical that all of the dotnet mappings continue working as they have before, if any of that changes existing packages/projects will be break.

Since netstandard will include PCL mappings the netstandard and dotnet TxMs are not equivalent. I think you'll need to leave the current dotnet mappings and add additional mappings for netstandard, and then netstandard to dotnet mappings.

@davidfowl  A dotnet project shouldn't be able to install a netstandard library, right? Or is that allowed?

emgarten commented:
You'll need tests covering
- GetNearest for netstandard, see: FrameworkReducer_GetNearestWithGenerations and similar tests
- Compatibility tests for netstandard -> dotnet
- Negative compatibility tests where dotnet cannot install netstandard and the other way around
- Tests for future versions of netstandard, netstandard0.0 and other possible variations to make sure they behave as expected

davidfowl commented:
> @davidfowl A dotnet project shouldn't be able to install a netstandard library, right? Or is that allowed?

We need to think through that transition story. We can choose to keep netstandard pure and make packages convert from the bottom up but it might be a bit rough in the beginning.

joelverhagen commented:
Addressed comments and ready for another look through. Per the `dotnet` / `netstandard` compatibility, I made `netstandard` support `dotnet` (e.g. `netstandard1.2` supports `dotnet5.3` and lower).

emgarten commented on src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs:
```diff
@@ -266,10 +302,14 @@ public IEnumerable<FrameworkSpecificMapping> EquivalentProfiles
+                                FrameworkConstants.FrameworkIdentifiers.Dnx),
+
+                            // NetStandard is a subset of DNXCore
+                            new KeyValuePair<string, string>(
```
This mapping is incorrect. It would say that netstandard5.5 is compatible with dnxcore5.5, basically every netstandard version would install into dnxcore. 

The dnxcore -> dotnet mapping needs to stay in place for backwards compatibility. Just add the old one back and leave netstandard out for this.

joelverhagen replied:
I believe this is fixed on [my latest iteration](https://github.com/NuGet/NuGet.Client/blob/jver-1691/src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs#L272) -- correct me if I am missing something.

emgarten commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -14,7 +14,13 @@ public class CompatibilityTests
 
-        // dnxcore50 -> dotnet
+        // dnxcore50 -> dotnet, netstandard
+        [InlineData("dnxcore50", "netstandard1.5", true)]
```
Is this allowed?

joelverhagen replied:
Per @ericstj's [change](https://github.com/NuGet/NuGet.Client/commit/a15e8ff0c6de85eb8da995672ce4bb1751fc141d#diff-466ac121bd03ff666f4f494225859d8eL324), `dnxcore50` supports `dotnet5.6`. We are saying that anything that supports `dotnet5.6` also supports `netstandard1.5`, I tested this.

emgarten replied:
:+1:

emgarten commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -196,7 +334,22 @@ public class CompatibilityTests
+        [InlineData("netstandard1.0", "dnxcore50", false)]
+        [InlineData("netstandard1.1", "dnxcore50", false)]
+        [InlineData("netstandard1.2", "dnxcore50", false)]
+        [InlineData("netstandard1.3", "dnxcore50", false)]
```
add netstandard -> dotnet also

emgarten commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/FrameworkComparerTests.cs:
```diff
@@ -55,6 +63,10 @@ public void FrameworkComparer_FrameworkOrderingWithPreferredAndNormalSorting()
+            Assert.Equal(fw11, list[10]);
+            Assert.Equal(fw12, list[11]);
+            Assert.Equal(fw13, list[12]);
+            Assert.Equal(fw14, list[13]);
```
this is incorrect, full frameworks should come before the packages based frameworks.

joelverhagen replied:
Agreed. This case is not possible with the reducer so I added the split precedence lists (as we spoke about offline).

emgarten commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -805,7 +810,12 @@ private void AddPortableOptionalFrameworks(IEnumerable<KeyValuePair<int, NuGetFr
         public void AddFrameworkPrecedenceMappings(IEnumerable<string> mappings)
         {
-            _frameworkPrecedence.AddRange(mappings);
+            int index = 0;
```
There could be multiple mappings that are passed in. The index will be reset between each one. This wasn't fully working before either, but now it looks like it could hit more conflicts.

joelverhagen replied:
Fixed.

emgarten commented on src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs:
```diff
@@ -478,6 +550,8 @@ public IEnumerable<string> FrameworkPrecedence
                         FrameworkConstants.FrameworkIdentifiers.NetCore,
                         FrameworkConstants.FrameworkIdentifiers.Windows,
                         FrameworkConstants.FrameworkIdentifiers.WindowsPhoneApp,
+                        FrameworkConstants.FrameworkIdentifiers.NetStandard,
```
I don't think this will work. Packages based platforms need to go last, this will put them ahead of all unlisted platforms.

joelverhagen replied:
Split the precedence lists, as agreed offline.

emgarten commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -803,9 +813,20 @@ private void AddPortableOptionalFrameworks(IEnumerable<KeyValuePair<int, NuGetFr
-            _frameworkPrecedence.AddRange(mappings);
+            if (mappings != null)
+            {
+                int index = destination.Count;
```
this could just use destination.Count everywhere instead of index

joelverhagen replied:
Fixed.

emgarten commented:
Looks great :shipit:
